src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/StaffConfigManager.kt

```diff
@@ -58,6 +69,138 @@ class StaffConfigManager(
     }
   }
 
+  fun setStaffDetails(
+    prisonCode: String,
+    staffId: Long,
+    request: StaffDetailsRequest,
+  ) {
+    val policy = AllocationContext.get().policy
+    val prisonConfig = prisonConfigRepository.findByCode(prisonCode) ?: PrisonConfiguration.default(prisonCode)
+    val staffConfig = staffConfigRepository.findByStaffId(staffId)
+    val staffRole: Pair<StaffRole?, NomisStaffRole?>? =
+      if (request.staffRole.isPresent) {
+        if (policy.nomisUseRoleCode != null) {
+          Pair(null, prisonApi.getKeyworkerForPrison(prisonCode, staffId))
+        } else {
+          Pair(staffRoleRepository.findByPrisonCodeAndStaffId(prisonCode, staffId), null)
+        }
+      } else {
+        null
+      }
+
+    val createConfig = staffConfig == null
+    val createRole =
+      request.staffRole.isPresent &&
+        request.staffRole.get() != null &&
+        staffRole!!.first == null &&
+        staffRole.second == null
+    val patchRole = request.staffRole.isPresent && request.staffRole.get() != null && !createRole
+    val deactivateJobRoleAndConfig = request.staffRole.isPresent && request.staffRole.get() == null
+    val removeActiveAllocations = request.deactivateActiveAllocations.orElse(false) || deactivateJobRoleAndConfig
+
+    if (createRole && !request.staffRole.get()!!.isValidToCreate()) {
+      throw ResponseStatusException(HttpStatusCode.valueOf(400), "Insufficient parameters to create staff job role")
+    }
+
+    if (createConfig) {
+      request.apply {
+        staffConfigRepository.save(
+          StaffConfiguration(
+            referenceDataRepository.getReferenceData(
+              ReferenceDataKey(ReferenceDataDomain.STAFF_STATUS, if (status.isPresent) status.get() else "ACTIVE"),
+            ),
+            capacity.orElse(prisonConfig.capacity),
+            allowAutoAllocation.orElse(true),
+            reactivateOn.orElse(null),
+            staffId,
+          ),
+        )
+      }
+    } else if (!deactivateJobRoleAndConfig) {
+      staffConfig.patch(request)
+    }
+
+    if (removeActiveAllocations) {
```

Comment:
Similar to the previous implementation, I think it would be better to deallocate last in case the call to the NOMIS User and Roles API fails